File: java/refactoring-methods/01-extract-method.java
import java.util.Map;

/**
 * 1. Method Extraction (Extract Method)
 *
 * BEFORE: A method contains too much logic, making it hard to understand
 */
class OrderProcessorBefore {
    public double processOrder(Map<String, Object> order) throws Exception {
        // Validate order
        if ((Double) order.get("total") <= 0) {
            throw new Exception("Invalid order total");
        }

        // Calculate tax
        double tax = (Double) order.get("subtotal") * 0.08;

        // Calculate shipping
        double shipping = (Double) order.get("weight") > 10 ? 15.00 : 5.00;

        // Calculate total
        double total = (Double) order.get("subtotal") + tax + shipping;

        // Save to database
        saveOrder(order, total);

        return total;
    }

    private void saveOrder(Map<String, Object> order, double total) {
        // Database save logic
    }
}

/**
 * AFTER: Extract methods to separate concerns
 */
class OrderProcessorAfter {
    public double processOrder(Map<String, Object> order) throws Exception {
        validateOrder(order);

        double tax = calculateTax(order);
        double shipping = calculateShipping(order);
        double total = calculateTotal(order, tax, shipping);

        saveOrder(order, total);

        return total;
    }

    private void validateOrder(Map<String, Object> order) throws Exception {
        if ((Double) order.get("total") <= 0) {
            throw new Exception("Invalid order total");
        }
    }

    private double calculateTax(Map<String, Object> order) {
        return (Double) order.get("subtotal") * 0.08;
    }

    private double calculateShipping(Map<String, Object> order) {
        return (Double) order.get("weight") > 10 ? 15.00 : 5.00;
    }

    private double calculateTotal(Map<String, Object> order, double tax, double shipping) {
        return (Double) order.get("subtotal") + tax + shipping;
    }

    private void saveOrder(Map<String, Object> order, double total) {
        // Database save logic
    }
}

/**
 * 2. Embedding a method (Inline Method)
 *
 * BEFORE: A method is too simple and adds no value
 */
class UserBefore {
    private String firstName;
    private String lastName;

    public String getFullName() {
        return getFirstName() + " " + getLastName();
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }
}

/**
 * AFTER: Inline the simple method
 */
class UserAfter {
    private String firstName;
    private String lastName;

    public String getFullName() {
        return firstName + " " + lastName;
    }

    // getFirstName() and getLastName() methods removed
}

File: php/refactoring-methods/01-extract-method.php
<?php

/**
 * 1. Method Extraction (Extract Method)
 *
 * BEFORE: A method contains too much logic, making it hard to understand
 */
class OrderProcessorBefore {
    public function processOrder($order) {
        // Validate order
        if ($order['total'] <= 0) {
            throw new Exception('Invalid order total');
        }

        // Calculate tax
        $tax = $order['subtotal'] * 0.08;

        // Calculate shipping
        $shipping = $order['weight'] > 10 ? 15.00 : 5.00;

        // Calculate total
        $total = $order['subtotal'] + $tax + $shipping;

        // Save to database
        $this->saveOrder($order, $total);

        return $total;
    }

    private function saveOrder($order, $total) {
        // Database save logic
    }
}

/**
 * AFTER: Extract methods to separate concerns
 */
class OrderProcessorAfter {
    public function processOrder($order) {
        $this->validateOrder($order);

        $tax = $this->calculateTax($order);
        $shipping = $this->calculateShipping($order);
        $total = $this->calculateTotal($order, $tax, $shipping);

        $this->saveOrder($order, $total);

        return $total;
    }

    private function validateOrder($order) {
        if ($order['total'] <= 0) {
            throw new Exception('Invalid order total');
        }
    }

    private function calculateTax($order) {
        return $order['subtotal'] * 0.08;
    }

    private function calculateShipping($order) {
        return $order['weight'] > 10 ? 15.00 : 5.00;
    }

    private function calculateTotal($order, $tax, $shipping) {
        return $order['subtotal'] + $tax + $shipping;
    }

    private function saveOrder($order, $total) {
        // Database save logic
    }
}

/**
 * 2. Embedding a method (Inline Method)
 *
 * BEFORE: A method is too simple and adds no value
 */
class UserBefore {
    public function getFullName() {
        return $this->getFirstName() . ' ' . $this->getLastName();
    }

    public function getFirstName() {
        return $this->firstName;
    }

    public function getLastName() {
        return $this->lastName;
    }
}

/**
 * AFTER: Inline the simple method
 */
class UserAfter {
    public function getFullName() {
        return $this->firstName . ' ' . $this->lastName;
    }

    // getFirstName() and getLastName() methods removed
}
